log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/generator/internal/TypeLookup.java

```diff
@@ -42,18 +44,106 @@ private TypeLookup(final Iterable<? extends PluginSet> pluginSets, final Predica
 
     private void mergeDescriptors(Iterable<? extends PluginSet> pluginSets) {
         pluginSets.forEach(pluginSet -> {
-            pluginSet.getScalars().forEach(scalar -> {
-                final ArtifactSourcedType sourcedType = ArtifactSourcedType.ofPluginSet(pluginSet, scalar);
-                putIfAbsent(scalar.getClassName(), sourcedType);
-            });
-            pluginSet.getAbstractTypes().forEach(abstractType -> {
-                final ArtifactSourcedType sourcedType = ArtifactSourcedType.ofPluginSet(pluginSet, abstractType);
-                putIfAbsent(abstractType.getClassName(), sourcedType);
-            });
-            pluginSet.getPlugins().forEach(pluginType -> {
-                final ArtifactSourcedType sourcedType = ArtifactSourcedType.ofPluginSet(pluginSet, pluginType);
-                putIfAbsent(pluginType.getClassName(), sourcedType);
-            });
+            mergeScalarTypes(pluginSet);
+            mergeAbstractTypes(pluginSet);
+            mergePluginTypes(pluginSet);
```

Comment:
What about using [`Map#merge`](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#merge-K-V-java.util.function.BiFunction-) with three different merging functions?

Reply:
Implemented in 2cfeeb64cccb7f60d018c4cf8a4056f437c850ba. I think it increased the cognitive load, but anyway...